manager/allocator/network/driver/driver.go

```diff
@@ -0,0 +1,259 @@
+	}
+	// the use of getDriver here is kind of redundant, because we would expect
+	// the caller to call IsNetworkNodeLocal first, which will do the same
+	// retry dance, but we'll leave this in for safety's sake.
+	driver, caps, err := a.getDriver(driverName)
+	if err != nil {
+		return err
+	}
+	// No swarm-level allocation can be provided by the network driver for
+	// node-local networks. Only thing needed is populating the driver's name
+	// in the driver's state.
+	if caps.DataScope == datastore.LocalScope {
+		n.DriverState = &api.Driver{
+			Name: driverName,
+		}
+		// In order to support backward compatibility with older daemon
+		// versions which assumes the network attachment to contains
+		// non nil IPAM attribute, passing an empty object
+		n.IPAM = &api.IPAMOptions{Driver: &api.Driver{}}
+		return nil
+	}
+
+	var options map[string]string
+	// NOTE(dperny): if we supported network updates, we would have to
+	// reconcile the driver state options in the spec with those in the object.
+	// However, since network updates are not supported, we are not doing that.
+	if n.Spec.DriverConfig != nil {
+		options = n.Spec.DriverConfig.Options
+	}
+
+	ipamData, err := getIpamData(n)
+	if err != nil {
+		return err
+	}
+	// now we have everything we need and can call NetworkAllocate. Note that
+	// the last param is nil. That param is IPv6 data, but we do not support
+	// IPv6
+	ds, err := driver.NetworkAllocate(n.ID, options, ipamData, nil)
+	if err != nil {
+		return errors.ErrInternal("driver.NetworkAllocate call failed: %v", err)
+	}
+
+	// finally, update the network object with the driver state we've obtained,
+	// and we'll be done and ready to commit
+	n.DriverState = &api.Driver{
+		Name:    driverName,
+		Options: ds,
+	}
+	return nil
+}
+
+// Deallocate frees the network driver state.
+func (a *allocator) Deallocate(n *api.Network) error {
+	// if the driver state is already nil, we have nothing to do because it was
+	// never allocated
+	if n.DriverState == nil {
+		return nil
+	}
+	driver, caps, err := a.getDriver(n.DriverState.Name)
+	if err != nil {
+		return err
+	}
+	// if the driver is locally scoped, we also have nothing to do here
+	if caps.DataScope == datastore.LocalScope {
+		return nil
+	}
+	return driver.NetworkFree(n.ID)
+}
+
+// IsNetworkNodeLocal is a helper function that gets the network driver and
+// checks its capabilities to see if it is a node-local network.
+//
+// There is a weird cross-dependency of the IPAM and Driver allocators. The
+// IPAM allocator cannot be run on node-local netowrks, but the Driver
+// allocator needs IPAM data filled in in order to allocate the network. To
+// avoid this, we expose this method, which just returns true if the network is
+// node-local, indicating that the IPAM allocator should not be run.
+func (a *allocator) IsNetworkNodeLocal(n *api.Network) (bool, error) {
+	name := DefaultDriver
+	if n.DriverState != nil && n.DriverState.Name != "" {
```

Comment:
Is it possible to have a `n.DriverState` but not a `n.Spec.DriverConfig`?

Reply:
It is possible to have no DriverConfig on the spec; that field can be nil, in which case, the defaults will be filled in. We could say if there was no driver config, because we know the default to be `overlay`, a global scoped network, that we could return `false, nil`, but I think it's cleaner to do it this way.